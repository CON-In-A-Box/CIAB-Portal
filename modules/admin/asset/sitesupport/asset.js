/* jshint browser: true */
/* jshint -W097 */
/* globals  basicBackendRequest, apiRequest, showSpinner, hideSpinner */
/* exported  uploadAsset, startAssetUpload, populateAssetConfiguration,
             setAssetField */

'use strict';

function uploadAsset(id, file) {
  var data = file[0];
  var formData = new FormData();
  formData.append('AssetID', id);
  formData.append('AssetData', data);
  basicBackendRequest('POST', 'admin/asset', formData,
    function() {
      location.reload();
    });
}

function startAssetUpload(asset) {
  var upload = document.getElementById('assetToUpload.' + asset);
  upload.click();
}

function setAssetField(field) {
  var value = document.getElementById('config_' + field).value;
  showSpinner();
  apiRequest('PUT','admin/configuration',
    'Field=' + field + '&Value=' + value)
    .then(function() {
      hideSpinner();
    })
    .catch(function(response) {
      hideSpinner();
      if (response instanceof Error) { throw response; }
    });
}

function createTableCell() {
  var cell = document.createElement('DIV');
  cell.classList.add('UI-table-cell');
  cell.classList.add('UI-padding');
  return cell;
}

function populateAssetConfiguration() {
  basicBackendRequest('GET', 'admin/asset', 'configuration=1',
    function(request) {
      var base = document.getElementById('config_table');
      base.innerHTML = '';
      var output = JSON.parse(request.responseText);
      for (var property in output) {
        var line = document.createElement('DIV');
        line.classList.add('UI-table-row');
        line.classList.add('UI-padding');
        var value = createTableCell();
        value.innerHTML = property;
        line.appendChild(value);
        value = createTableCell();
        value.innerHTML = output[property].description;
        line.appendChild(value);

        if (output[property].source == 'ENV' || output[property].readonly) {
          value = createTableCell();
          value.innerHTML = output[property].value;
          line.appendChild(value);

          if (!output[property].readonly) {
            value = createTableCell();
            value.innerHTML = '&lt;ENVIRONMENT&gt;';
            line.appendChild(value);
          } else {
            value = createTableCell();
            value.innerHTML = '';
            line.appendChild(value);
          }
        } else {
          value = createTableCell();
          var edit = document.createElement('INPUT');
          edit.setAttribute('type', 'text');
          edit.id = 'config_' + property;
          edit.value = output[property].value;
          edit.classList.add('UI-input');
          value.appendChild(edit);
          line.appendChild(value);

          value = createTableCell();
          var button = document.createElement('BUTTON');
          button.classList.add('UI-eventbutton');
          button.setAttribute('onclick', 'setAssetField("' + property + '");');
          button.innerHTML = 'Set';
          value.appendChild(button);
          line.appendChild(value);
        }

        base.appendChild(line);

        if (property == 'ASSET_TYPES') {
          var upload = document.getElementById('assetToUpload.org-icon');
          upload.setAttribute('accept', output[property].value);
        }
      }
    });

}
