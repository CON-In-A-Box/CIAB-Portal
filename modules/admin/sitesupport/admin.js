/* jshint browser: true */
/* jshint -W097 */
/* globals apiRequest, confirmbox, basicBackendRequest, showSpinner,
           hideSpinner, alertbox */
/* exported downloadLog, setField, addField, removeAdmin, updateMemberships,
            doSUDO, rebuildSCSS, setPassword */

function downloadLog() {
  window.location = 'index.php?Function=admin&downloadLog=db';
}

function basicAdminRequest(parameter, finish) {
  basicBackendRequest('POST', 'admin', parameter, finish);
}

function setField(field) {
  var value = document.getElementById('config_' + field).value;
  value = btoa(value);
  basicAdminRequest('&setField=' + field + '&value=' + value, function() {
    location.reload();
  });
}

function addField() {
  var field = document.getElementById('new_Field').value;
  var value = document.getElementById('new_Value').value;
  value = btoa(value);
  basicAdminRequest('&newField=' + field + '&value=' + value, function() {
    location.reload();
  });
}

function removeAdmin() {
  basicAdminRequest('&removeAdmin=true', function() {
    location.reload();
  });
}

function updateMemberships() {
  confirmbox('Update Memberships?', 'This operations will run in the ' +
    'background and can take several minutes to complete.' +
    '<br><em>Run Command?</em>').then(function() {
    basicAdminRequest('&updateMemberships=true', function() {});
  });
}

function doSUDO() {
  var value = document.getElementById('sudo_id').value;
  apiRequest('POST', 'admin/SUDO/' + value, '');
}

function rebuildSCSS() {
  basicAdminRequest('&rebuildSCSS=true', function() {
    location.reload();
  });
}

function setPassword() {
  var newPassword = document.getElementById('tmp_passwd').value;
  var account = document.getElementById('tmp_login').value;

  showSpinner();
  apiRequest('PUT', 'member/' + account + '/password' ,
    'NewPassword=' + newPassword)
    .then(function() {
      hideSpinner();
      alertbox('Password Updated');
    })
    .catch(function(response) {
      hideSpinner();
      alertbox('Password Reset Failed: ' + response.responseText);
    });
}
