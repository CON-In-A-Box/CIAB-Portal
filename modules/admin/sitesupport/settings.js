/*
 * Javacript for the Site Configuration page
 */

/* jshint browser: true */
/* jshint -W097 */
/* globals settingsTable */

function load() {
  new settingsTable({api: 'admin/configuration'}).createElement();
}

if (window.addEventListener) {
  window.addEventListener('load', load);
} else {
  window.attachEvent('onload', load);
}
