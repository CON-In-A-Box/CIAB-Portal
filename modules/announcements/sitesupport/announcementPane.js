/*
 * Javacript for the Announcements Main Panel
 */

/* jshint browser: true */
/* jshint -W097 */
/* globals apiRequest */

var announcementPane = (function(options) {
  'use strict';

  var settings = Object.assign(
    {
    }, options);

  return {
    options: function(opts) {
      settings = Object.assign(settings, opts);
    },

    displayAnnouncements: function(frame) {
      var target = document.getElementById('announcement_pane');
      if (frame !== null) {
        target.appendChild(frame[0]);
      }
    },

    addAnnouncement: function(table, data2) {
      var line = document.createElement('DIV');
      line.classList.add('UI-table-row');
      line.classList.add('UI-white');
      var f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.classList.add('UI-announcement');
      f.id = 'announcement-' + data2.id + '-note';

      var a = document.createElement('DIV');
      a.appendChild(document.createTextNode(data2.text));
      var s = document.createElement('SPAN');
      s.classList.add('UI-announcement-source');
      s.appendChild(document.createTextNode('  -  ' + data2.departmentId.name));
      a.appendChild(s);
      s.appendChild(document.createTextNode(' (' + data2.postedOn + ')'));
      a.appendChild(s);
      f.appendChild(a);
      line.appendChild(f);
      table.appendChild(line);
      return 0;
    },

    emptyAnnouncementBlock: function() {
      var block = document.createElement('DIV');
      block.id = 'announcement-block';
      block.classList.add('UI-container');
      block.classList.add('event-color-primary');
      var title = document.createElement('div');
      title.classList.add('event-color-primary');
      title.classList.add('UI-center');
      var txt = document.createElement('B');
      txt.classList.add('UI-bold');
      txt.style.fontSize = '125%';
      txt.appendChild(document.createTextNode('Announcements'));
      title.appendChild(txt);
      block.appendChild(title);
      var contents = document.createElement('DIV');
      contents.classList.add('UI-container');
      block.appendChild(contents);
      var table = document.createElement('DIV');
      table.classList.add('UI-table');
      table.classList.add('UI-table-padded');
      contents.appendChild(table);
      return [block, table];
    },

    buildAnnouncementBlock: function(frame, result, data) {
      var rc = frame[0];
      var table = rc[1];
      return announcementPane.addAnnouncement(table, data);
    },

    load: function() {
      var frame = [ announcementPane.emptyAnnouncementBlock() ];
      apiRequest('GET',
        'member/current/announcements',
        'maxResults=all&include=departmentId')
        .then(function(response) {
          var target = document.getElementById('announcement_pane');
          var result = JSON.parse(response.responseText);
          if (result.data.length > 0) {
            result.data.forEach(function(data) {
              announcementPane.buildAnnouncementBlock(frame, result, data);
            });
            var line = document.createElement('DIV');
            line.classList.add('event-color-primary');
            line.appendChild(document.createTextNode('\u00A0'));
            frame[0][0].appendChild(line);
            target.classList.remove('UI-hide');
          } else {
            target.classList.add('UI-hide');
          }
        });
      announcementPane.displayAnnouncements(frame[0]);
    }
  };
})();

if (window.addEventListener) {
  window.addEventListener('load', announcementPane.load);
} else {
  window.attachEvent('onload', announcementPane.load);
}
