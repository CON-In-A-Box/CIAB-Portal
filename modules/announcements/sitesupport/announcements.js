/*
 * Javacript for the Announcements page
 */

/* jshint browser: true */
/* jshint -W097 */
/* globals apiRequest, showSpinner, hideSpinner, showSidebar,
           confirmbox, alertbox, departmentDropdown */

var announcementPage = (function(options) {
  'use strict';

  var permissions = null;

  var settings = Object.assign(
    {
    }, options);

  return {
    options: function(opts) {
      settings = Object.assign(settings, opts);
    },

    openEdit: function() {
      var dataset = this.dataset;
      var id = dataset.ciabAnnouncementId;
      var text = document.getElementById('announcement-' + id + '-text')
        .innerHTML;
      document.getElementById('announcement_text').value = text;
      document.getElementById('department_dropdown_select')
        .value = dataset.ciabAnnouncementDepartment;

      var canEdit = Object.prototype.hasOwnProperty.call(permissions,
        dataset.ciabAnnouncementDepartment + '_announcement_post');
      var canDelete = Object.prototype.hasOwnProperty.call(permissions,
        dataset.ciabAnnouncementDepartment + '_announcement_delete');

      var button = document.getElementById('remove_event_button');
      if (canDelete) {
        button.classList.remove('UI-hide');
        button.dataset.ciabAnnouncementId = id;
      } else {
        button.classList.add('UI-hide');
      }
      button = document.getElementById('save_event_button');
      if (canEdit) {
        button.classList.remove('UI-hide');
        button.dataset.ciabAnnouncementId = id;
      } else {
        button.classList.add('UI-hide');
      }

      document.getElementById('modify_title').innerHTML = 'Modify';
      showSidebar('modify_announcement');
    },

    removeEvent: function(button) {
      var dataset = button.dataset;
      var id = dataset.ciabAnnouncementId;
      confirmbox('Remove Announcement', 'Confirm removal of this announcement')
        .then(
          function() {
            showSpinner();
            apiRequest('DELETE', 'announcement/' + id, '')
              .then(function() {
                location.reload();
              })
              .catch(function(response) {
                if (response instanceof Error) { throw response; }
                var data = JSON.parse(response.responseText);
                hideSpinner();
                alertbox('Delete Failed', data.status);
              });
          });
    },

    updateEvent: function(button) {
      var dataset = button.dataset;
      var id = dataset.ciabAnnouncementId;
      var text = document.getElementById('announcement_text').value;
      var dept = document.getElementById('department_dropdown_select').value;
      var scope = document.getElementById('scope_drop').value;
      var email = (document.getElementById('announcement_email').checked ?
        1 : 0);
      var message = 'Confirm update of this announcement';
      if (email && id == -1) {
        if (scope == '0') {
          message += '<div class="UI-border UI-margin UI-red"> ' +
          '<span class="UI-bold">' +
          '<em>WARNING</em>: This will also email EVERY member that has EVER ' +
          'registered for the event!</span></div>';
        }
        else if (scope == '1') {
          message += '<div class="UI-border UI-margin UI-yellow"> ' +
          '<span class="UI-bold">This will also email ' +
          'EVERY member of the event staff</span>';
        }
        else if (scope == '2') {
          var d = document.getElementById('department_dropdown_select');
          message += '<div class="UI-border UI-margin"> ' +
          '<span>This will also email every member of the \'' +
          d.options[d.selectedIndex].text + '\' Department</span>';
        }
      }
      confirmbox('Update Announcement', message)
        .then(
          function() {
            showSpinner();
            var method = 'PUT';
            if (id == -1) {
              method = 'POST';
              id = dept;
            }
            apiRequest(method, 'announcement/' + id,
              'Scope=' + scope + '&Text=' + encodeURI(text) + '&Department=' +
              dept + '&Email=' + email)
              .then(function() {
                location.reload();
              })
              .catch(function(response) {
                if (response instanceof Error) { throw response; }
                var data = JSON.parse(response.responseText);
                hideSpinner();
                alertbox('Update Failed', data.status);
              });
          })
        .catch(function(response) {
          if (response instanceof Error) { throw response; }
        });
    },

    displayAnnouncements: function(cache) {
      var target = document.getElementById('headline_section');
      cache.forEach(function(data) {
        if (data !== null) {
          target.appendChild(data[0]);
        }
      });
    },

    addAnnouncement: function(table, data2) {
      var line = document.createElement('DIV');
      line.classList.add('UI-table-row');

      var f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.id = 'announcement-' + data2.id + '-text';
      f.appendChild(document.createTextNode(data2.text));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.id = 'announcement-' + data2.id + '-posted';
      f.appendChild(document.createTextNode(data2.postedOn));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.id = 'announcement-' + data2.id + '-poster';
      var name = data2.postedBy.firstName + ' ' + data2.postedBy.lastName +
        ' (' + data2.postedBy.email + ')';
      f.appendChild(document.createTextNode(name));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.id = 'announcement-' + data2.id + '-scope';
      if (data2.scope >= 2)
      {
        f.appendChild(document.createTextNode('Department Only'));
      }
      else if (data2.scope >= 1)
      {
        f.appendChild(document.createTextNode('All ConCom'));
      }
      else
      {
        f.appendChild(document.createTextNode('Convention Wide'));
      }
      line.appendChild(f);

      f = document.createElement('DIV');
      f.setAttribute('name', 'announcement-table-modify-' +
        data2.departmentId.id);
      f.classList.add('UI-table-cell');
      f.classList.add('UI-center');
      f.classList.add('UI-hide');
      var button = document.createElement('button');
      button.innerHTML = '<i class="fas fa-calendar-check"></i>';
      button.dataset.ciabAnnouncementId = data2.id;
      button.dataset.ciabAnnouncementDepartment = data2.departmentId.id;
      button.onclick = announcementPage.openEdit;
      button.classList.add('UI-button');
      f.appendChild(button);
      line.appendChild(f);
      table.appendChild(line);
    },

    generateTableHeader: function(dept) {
      var line = document.createElement('DIV');
      line.id = 'announcement-table-header-' + dept.id;
      line.classList.add('UI-table-row');

      var f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.appendChild(document.createTextNode('Announcement'));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.appendChild(document.createTextNode('Posted On'));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.appendChild(document.createTextNode('Posted By'));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.classList.add('UI-table-cell');
      f.appendChild(document.createTextNode('Scope'));
      line.appendChild(f);

      f = document.createElement('DIV');
      f.setAttribute('name', 'announcement-table-modify-' + dept.id);
      f.classList.add('UI-table-cell');
      f.classList.add('UI-hide');
      f.classList.add('UI-center');
      f.appendChild(document.createTextNode(''));
      line.appendChild(f);
      return line;
    },

    newAnnouncement: function() {
      var dataset = this.dataset;
      document.getElementById('announcement_text').value = '';
      document.getElementById('announcement_email').checked = false;
      if (dataset.ciabAnnouncementDepartment) {
        document.getElementById('department_dropdown_select')
          .value = dataset.ciabAnnouncementDepartment;
      } else {
        document.getElementById('department_dropdown_select')
          .selectedIndex = 0;
      }
      var button = document.getElementById('remove_event_button');
      button.classList.add('UI-hide');
      button = document.getElementById('save_event_button');
      button.classList.remove('UI-hide');
      button.dataset.ciabAnnouncementId = -1;

      document.getElementById('modify_title').innerHTML = 'Add';
      showSidebar('modify_announcement');
    },

    emptyAnnouncementBlock: function(dept) {
      var block = document.createElement('DIV');
      block.id = 'announcement-block-' + dept.id;
      block.classList.add('UI-container');
      var title = document.createElement('H2');
      title.classList.add('UI-secondary-sectionbar');
      title.appendChild(document.createTextNode(dept.name));
      var button = document.createElement('button');
      button.innerHTML = '<i class="fas fa-calendar-plus"></i>';
      title.appendChild(button);
      button.classList.add('UI-button');
      button.classList.add('UI-right');
      button.classList.add('UI-hide');
      button.dataset.ciabAnnouncementDepartment = dept.id;
      button.onclick = announcementPage.newAnnouncement;
      block.appendChild(title);
      var contents = document.createElement('DIV');
      contents.classList.add('UI-container');
      block.appendChild(contents);
      var table = document.createElement('DIV');
      table.classList.add('UI-table-all');
      contents.appendChild(table);
      var line = announcementPage.generateTableHeader(dept);
      table.appendChild(line);
      return [block, table];
    },

    buildAnnouncementBlock: function(cache, result, data) {
      var table;
      var rc;
      var dept = cache[parseInt(data.departmentId.id)];
      if (!dept) {
        rc = announcementPage.emptyAnnouncementBlock(data.departmentId);
        table = rc[1];
        cache[parseInt(data.departmentId.id)] = rc;
      } else {
        rc = cache[parseInt(data.departmentId.id)];
        table = rc[1];
      }
      announcementPage.addAnnouncement(table, data);
    },

    filter: function(dept) {
      var id = dept.id.id.toString() + '_announcement_post';
      return Object.prototype.hasOwnProperty.call(permissions, id);
    },

    loadAnnouncements: function() {
      showSpinner();
      apiRequest('GET',
        'member/current/announcements',
        'maxResults=all&include=departmentId,postedBy')
        .then(function(response) {
          var result = JSON.parse(response.responseText);
          if (result.data.length > 0) {
            var cache = Array();
            result.data.forEach(function(data) {
              announcementPage.buildAnnouncementBlock(cache, result, data);
            });
            announcementPage.displayAnnouncements(cache);
          }
          apiRequest('GET',
            'permissions/method/announcement',
            'maxResults=all')
            .then(function(response) {
              result = JSON.parse(response.responseText);
              var havePost = false;
              permissions = {};
              if (result.data.length > 0) {
                result.data.forEach(function(data) {
                  if (data.allowed) {
                    permissions[data.subdata.departmentId + '_' +
                                  data.subtype] = data;
                    if (data.subtype == 'announcement_post') {
                      var sect = 'announcement-block-' +
                                 data.subdata.departmentId;
                      var block = document.getElementById(sect);
                      if (block !== null) {
                        var button = block.getElementsByTagName('button');
                        if (button !== null) {
                          button[0].classList.remove('UI-hide');
                        }
                      }
                      if (!havePost) {
                        havePost = true;
                        var add = document.getElementById(
                          'announcement-sectionbar-add');
                        add.classList.remove('UI-hide');
                        add.onclick = announcementPage.newAnnouncement;
                      }
                    }
                    if (data.subtype == 'announcement_delete' ||
                          data.subtype == 'announcement_put') {
                      var line = 'announcement-table-modify-' +
                            data.subdata.departmentId;
                      var cells = document.getElementsByName(line);
                      cells.forEach(function(cell) {
                        cell.classList.remove('UI-hide');
                      });
                    }
                  }
                });
              }
              departmentDropdown.build(
                {
                  div: 'dept_drop',
                  filter: announcementPage.filter
                }
              );
              hideSpinner();
            })
            .catch(function(response) {
              if (response instanceof Error) { throw response; }
              hideSpinner();
            });
        })
        .catch(function(response) {
          if (response instanceof Error) { throw response; }
          var target = document.getElementById('headline_section');
          target.innerHTML = response.responseText;
          hideSpinner();
        });
    }
  };
})();

if (window.addEventListener) {
  window.addEventListener('load', announcementPage.loadAnnouncements);
} else {
  window.attachEvent('onload', announcementPage.loadAnnouncements);
}
