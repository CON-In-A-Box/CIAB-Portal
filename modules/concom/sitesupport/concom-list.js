/* globals Vue, showSpinner, hideSpinner, Vuetify */
import staffListComponent from './components/staff-list.js';
import sectionNavComponent from './components/staff-section-nav.js';
import staffDivisionComponent from './components/staff-division.js';
import departmentHeaderComponent from './components/department-header.js';
import departmentMemberComponent from './components/department-member.js';
import staffDepartmentComponent from './components/staff-department.js';
import staffSidebarComponent from './components/staff-sidebar.js';
import lookupuser from '../../../sitesupport/vue/lookupuser.js';
import staffDivisionVisualComponent from './components/staff-division-visual.js';

function updateLoading() {
  this.isLoading = !this.isLoading;
  this.isLoading ? showSpinner() : hideSpinner();
}

const staffStandardTheme = {
  dark: false,
  colors: {
    primary: '#620272',
    secondary: '#00e500',
    error: '#B00020',
    info: '#2196F3',
    success: '#4CAF50',
    warning: '#FB8C00'
  }
}

const vuetify = Vuetify.createVuetify({
  theme: {
    defaultTheme: 'staffStandardTheme',
    themes: {
      staffStandardTheme
    }
  },
  defaults: {
    VBtn: {
      color: 'primary'
    },
    VSelect: {
      variant: 'outlined'
    },
    VListItem: {
      'color': 'white'
    }
  }
});

const staffApp = Vue.createApp({
  setup() {
    const isLoading = Vue.ref(false);
    Vue.provide('isLoading', {
      isLoading,
      updateLoading
    });

    return {
      isLoading
    }
  },
  methods: {
    updateLoading
  }
}).use(vuetify);

staffApp.component('staff-division', staffDivisionComponent);
staffApp.component('staff-section-nav', sectionNavComponent);
staffApp.component('staff-list', staffListComponent);
staffApp.component('department-header', departmentHeaderComponent);
staffApp.component('department-member', departmentMemberComponent);
staffApp.component('staff-department', staffDepartmentComponent);
staffApp.component('staff-sidebar', staffSidebarComponent);
staffApp.component('lookup-user', lookupuser);
staffApp.component('org-donut', staffDivisionVisualComponent);

staffApp.mount('#staff-app');

export default staffApp;
