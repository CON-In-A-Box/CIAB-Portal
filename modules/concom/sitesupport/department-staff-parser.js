const extractStaff = (staff) => {
  return {
    id: parseInt(staff.member?.id) ?? -1,
    deptStaffId: parseInt(staff.id) ?? -1,
    deptId: parseInt(staff.department.id) ?? -1,
    departmentName: staff.department?.name ?? '',
    divisionName: staff.department?.parent?.name,
    firstName: staff.member?.first_name ?? '',
    lastName: staff.member?.last_name ?? '',
    pronouns: staff.member?.pronouns ?? '',
    position: staff.position,
    email: staff.member?.email ?? '',
    note: staff.note ?? ''
  }
};

export const sortStaffByPosition = (staff, otherStaff) => {
  if (staff.position === 'Head' && otherStaff.position !== 'Head') {
    return -1;
  } else if (staff.position === 'Sub-Head' && otherStaff.position !== 'Sub-Head') {
    return 0;
  }

  return 1;
};

export const extractDepartmentStaff = (staffData) => {
  return staffData.map((item) => extractStaff(item));
};
