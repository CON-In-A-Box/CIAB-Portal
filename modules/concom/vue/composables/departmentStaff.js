/* globals apiRequest, Vue */
import { extractDepartmentStaff } from '../../sitesupport/department-staff-parser.js'

export function useDepartmentStaff() {
  const departmentStaff = Vue.ref(null);
  const loadingDepartment = Vue.ref(false);
  const departmentError = Vue.ref(null);

  const departmentStaffPositions = Vue.ref(null);
  const loadingPositions = Vue.ref(false);
  const departmentPositionsError = Vue.ref(null);

  async function fetchDepartmentStaff(departmentId) {
    if (departmentId === -1) {
      // Don't need to call the API, department is new.
      return;
    }

    loadingDepartment.value = true;

    try {
      const departmentResponse = await apiRequest('GET', `department/${departmentId}/staff?subdepartments=1&max_results=all`);
      const departmentStaffData = JSON.parse(departmentResponse.responseText);

      departmentStaff.value = extractDepartmentStaff(departmentStaffData.data);
    } catch (error) {
      departmentError.value = error;
    } finally {
      loadingDepartment.value = false;
    }
  }

  async function fetchDepartmentStaffPositions() {
    loadingPositions.value = true;

    try {
      const positionResponse = await apiRequest('GET', 'staff/positions');
      const parsedData = JSON.parse(positionResponse.responseText);

      departmentStaffPositions.value = parsedData.data;
    } catch (error) {
      departmentPositionsError.value = error;
    } finally {
      loadingPositions.value = false;
    }
  }

  return {
    departmentStaff,
    loadingDepartment,
    departmentError,
    fetchDepartmentStaff,
    departmentStaffPositions,
    loadingPositions,
    departmentPositionsError,
    fetchDepartmentStaffPositions
  };
}
