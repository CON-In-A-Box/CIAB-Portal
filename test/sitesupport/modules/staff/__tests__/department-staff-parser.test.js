/* globals describe, it, expect */
import { extractDepartmentStaff } from '../../../../../modules/concom/sitesupport/department-staff-parser';
import divisionStaff from './division_staff.json';
import departmentStaff from './department_staff.json';

describe('Department Staff Parser', () => {
  it.each([
    {
      input: divisionStaff,
      name: 'Division Staff'
    },
    {
      input: departmentStaff,
      name: 'Department Staff'
    }
  ])('parses $name', ({ input }) => {
    const result = extractDepartmentStaff(input.data);
    expect(result).toMatchSnapshot();
  });
});
