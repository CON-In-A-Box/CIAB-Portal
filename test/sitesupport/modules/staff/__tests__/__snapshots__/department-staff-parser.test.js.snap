// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Department Staff Parser parses Department Staff 1`] = `
[
  {
    "departmentName": "Book Swap",
    "deptId": 105,
    "deptStaffId": 45,
    "divisionName": "Activities",
    "email": "copgoc@email.net",
    "firstName": "copgoc",
    "id": 1408,
    "lastName": "FuaDigwaw",
    "note": "",
    "position": "Head",
    "pronouns": "",
  },
  {
    "departmentName": "Book Swap",
    "deptId": 105,
    "deptStaffId": 46,
    "divisionName": "Activities",
    "email": "MifRowgac@email.net",
    "firstName": "MifRowgac",
    "id": 1093,
    "lastName": "PakRovKitSewMoj",
    "note": "",
    "position": "Sub-Head",
    "pronouns": "",
  },
  {
    "departmentName": "Book Swap",
    "deptId": 105,
    "deptStaffId": 47,
    "divisionName": "Activities",
    "email": "baunocKum@email.net",
    "firstName": "baunocKum",
    "id": 1264,
    "lastName": "CegFaisiklaw",
    "note": "",
    "position": "Specialist",
    "pronouns": "",
  },
]
`;

exports[`Department Staff Parser parses Division Staff 1`] = `
[
  {
    "departmentName": "Activities",
    "deptId": 1,
    "deptStaffId": 1,
    "divisionName": undefined,
    "email": "test-admin@example.com",
    "firstName": "Test",
    "id": 1000,
    "lastName": "Admin",
    "note": "",
    "position": "Head",
    "pronouns": "He/Him",
  },
  {
    "departmentName": "Activities",
    "deptId": 1,
    "deptStaffId": 2,
    "divisionName": undefined,
    "email": "pezgiwPuh@email.net",
    "firstName": "pezgiwPuh",
    "id": 1109,
    "lastName": "setvocCab",
    "note": "",
    "position": "Sub-Head",
    "pronouns": "",
  },
  {
    "departmentName": "Activities",
    "deptId": 1,
    "deptStaffId": 3,
    "divisionName": undefined,
    "email": "Wirtesvai@email.net",
    "firstName": "Wirtesvai",
    "id": 1490,
    "lastName": "bokGedfer",
    "note": "",
    "position": "Specialist",
    "pronouns": "",
  },
]
`;
